dingding alert panic

When alert message include JSON data, process be panic...

ERROR LOG:
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x771972]

goroutine 3320373 [running]:
github.com/grafana/grafana/pkg/components/simplejson.(*Json).MarshalJSON(0x0, 0x117860b, 0x1a, 0xc420ac26c0, 0x4, 0x4)
        /go/src/github.com/grafana/grafana/pkg/components/simplejson/simplejson.go:75 +0x22
github.com/grafana/grafana/pkg/services/alerting/notifiers.(*DingDingNotifier).Notify(0xc420833090, 0xc420585e00, 0xc420d99788, 0xc420d02a00)
        /go/src/github.com/grafana/grafana/pkg/services/alerting/notifiers/dingding.go:77 +0x63a
github.com/grafana/grafana/pkg/services/alerting.(*notificationService).sendNotifications.func1(0x8, 0x11c5ad8)
        /go/src/github.com/grafana/grafana/pkg/services/alerting/notifier.go:69 +0x38
github.com/grafana/grafana/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1(0xc420ac2600, 0xc420391c00)
        /go/src/github.com/grafana/grafana/vendor/golang.org/x/sync/errgroup/errgroup.go:58 +0x57
created by github.com/grafana/grafana/vendor/golang.org/x/sync/errgroup.(*Group).Go
        /go/src/github.com/grafana/grafana/vendor/golang.org/x/sync/errgroup/errgroup.go:66 +0x66

--- a/pkg/services/alerting/notifiers/dingding.go
+++ b/pkg/services/alerting/notifiers/dingding.go
@@ -70,6 +70,26 @@ func (dd *DingDingNotifier) Notify(evalContext *alerting.EvalContext) error {
 		messageURL = ""
 	}
 
+	body, err := dd.genBody(evalContext, messageURL)
+	if err != nil {
+		return err
+	}
+
+	cmd := &models.SendWebhookSync{
+		Url:  dd.URL,
+		Body: string(body),
+	}
+
+	if err := bus.DispatchCtx(evalContext.Ctx, cmd); err != nil {
+		dd.log.Error("Failed to send DingDing", "error", err, "dingding", dd.Name)
+		return err
+	}
+
+	return nil
+}
+
+func (dd *DingDingNotifier) genBody(evalContext *alerting.EvalContext, messageURL string) ([]byte, error) {
+
 	q := url.Values{
 		"pc_slide": {"false"},
 		"url":      {messageURL},
@@ -124,22 +144,8 @@ func (dd *DingDingNotifier) Notify(evalContext *alerting.EvalContext) error {
 
 	if err != nil {
 		dd.log.Error("Failed to create Json data", "error", err, "dingding", dd.Name)
+		return nil, err
 	}
 
-	body, err := bodyJSON.MarshalJSON()
-	if err != nil {
-		return err
-	}
-
-	cmd := &models.SendWebhookSync{
-		Url:  dd.URL,
-		Body: string(body),
-	}
-
-	if err := bus.DispatchCtx(evalContext.Ctx, cmd); err != nil {
-		dd.log.Error("Failed to send DingDing", "error", err, "dingding", dd.Name)
-		return err
-	}
-
-	return nil
+	return bodyJSON.MarshalJSON()
 }

--- a/pkg/services/alerting/notifiers/dingding.go
+++ b/pkg/services/alerting/notifiers/dingding.go
@@ -1,12 +1,11 @@
 package notifiers
 
 import (
+	"encoding/json"
 	"fmt"
 	"net/url"
-	"strings"
 
 	"github.com/grafana/grafana/pkg/bus"
-	"github.com/grafana/grafana/pkg/components/simplejson"
 	"github.com/grafana/grafana/pkg/infra/log"
 	"github.com/grafana/grafana/pkg/models"
 	"github.com/grafana/grafana/pkg/services/alerting"
@@ -112,40 +111,32 @@ func (dd *DingDingNotifier) genBody(evalContext *alerting.EvalContext, messageUR
 		message += fmt.Sprintf("\\n%2d. %s: %s", i+1, match.Metric, match.Value)
 	}
 
-	var bodyStr string
+	var bodyMsg map[string]interface{}
 	if dd.MsgType == "actionCard" {
 		// Embed the pic into the markdown directly because actionCard doesn't have a picUrl field
 		if picURL != "" {
 			message = "![](" + picURL + ")\\n\\n" + message
 		}
 
-		bodyStr = `{
+		bodyMsg = map[string]interface{}{
 			"msgtype": "actionCard",
-			"actionCard": {
-				"text": "` + strings.Replace(message, `"`, "'", -1) + `",
-				"title": "` + strings.Replace(title, `"`, "'", -1) + `",
+			"actionCard": map[string]string{
+				"text":        message,
+				"title":       title,
 				"singleTitle": "More",
-				"singleURL": "` + messageURL + `"
-			}
-		}`
+				"singleURL":   messageURL,
+			},
+		}
 	} else {
-		bodyStr = `{
+		bodyMsg = map[string]interface{}{
 			"msgtype": "link",
-			"link": {
-				"text": "` + message + `",
-				"title": "` + title + `",
-				"picUrl": "` + picURL + `",
-				"messageUrl": "` + messageURL + `"
-			}
-		}`
+			"link": map[string]string{
+				"text":       message,
+				"title":      title,
+				"picUrl":     picURL,
+				"messageUrl": messageURL,
+			},
+		}
 	}
-
-	bodyJSON, err := simplejson.NewJson([]byte(bodyStr))
-
-	if err != nil {
-		dd.log.Error("Failed to create Json data", "error", err, "dingding", dd.Name)
-		return nil, err
-	}
-
-	return bodyJSON.MarshalJSON()
+	return json.Marshal(bodyMsg)
 }

--- a/pkg/services/alerting/notifiers/dingding_test.go
+++ b/pkg/services/alerting/notifiers/dingding_test.go
@@ -1,10 +1,12 @@
 package notifiers
 
 import (
+	"context"
 	"testing"
 
 	"github.com/grafana/grafana/pkg/components/simplejson"
 	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/alerting"
 	. "github.com/smartystreets/goconvey/convey"
 )
 
@@ -41,6 +43,16 @@ func TestDingDingNotifier(t *testing.T) {
 			So(notifier.Name, ShouldEqual, "dingding_testing")
 			So(notifier.Type, ShouldEqual, "dingding")
 			So(notifier.URL, ShouldEqual, "https://www.google.com")
+
+			Convey("genBody should not panic", func() {
+				evalContext := alerting.NewEvalContext(context.Background(),
+					&alerting.Rule{
+						State:   models.AlertStateAlerting,
+						Message: `{host="localhost"}`,
+					})
+				_, err = notifier.genBody(evalContext, "")
+				So(err, ShouldBeNil)
+			})
 		})
 	})
 }

